src/libpriv/rpmostree-core.c

```diff
@@ -2288,10 +2292,93 @@ on_async_import_done (GObject                    *obj,
 
   g_assert_cmpint (self->n_async_pkgs_imported, <, self->pkgs_to_import->len);
   self->n_async_pkgs_imported++;
+  self->n_async_running--;
   rpmostree_output_progress_n_items ("Importing", self->n_async_pkgs_imported,
                                      self->pkgs_to_import->len);
+  async_imports_mainctx_iter (self);
+}
+
+/* Queue an asynchronous import of a package */
+static gboolean
+import_one_package (RpmOstreeContext *self, DnfPackage *pkg,
+                    GCancellable *cancellable,
+                    GError **error)
+{
+  GVariant *rojig_xattrs = NULL;
+  if (self->rojig_pkg_to_xattrs)
+    {
+      rojig_xattrs = g_hash_table_lookup (self->rojig_pkg_to_xattrs, pkg);
+      if (!rojig_xattrs)
+        g_error ("Failed to find rojig xattrs for %s", dnf_package_get_nevra (pkg));
+    }
+
+  glnx_fd_close int fd = -1;
+  if (!rpmostree_context_consume_package (self, pkg, &fd, error))
+    return FALSE;
+
+  /* Only set SKIP_EXTRANEOUS for packages we know need it, so that
+   * people doing custom composes don't have files silently discarded.
+   * (This will also likely need to be configurable).
+   */
+  const char *pkg_name = dnf_package_get_name (pkg);
+
+  int flags = 0;
+  if (g_str_equal (pkg_name, "filesystem") ||
+      g_str_equal (pkg_name, "rootfiles"))
+    flags |= RPMOSTREE_IMPORTER_FLAGS_SKIP_EXTRANEOUS;
+
+  { gboolean docs;
+    g_assert (g_variant_dict_lookup (self->spec->dict, "documentation", "b", &docs));
+    if (!docs)
+      flags |= RPMOSTREE_IMPORTER_FLAGS_NODOCS;
+  }
+
+  /* TODO - tweak the unpacker flags for containers */
+  OstreeRepo *ostreerepo = get_pkgcache_repo (self);
+  g_autoptr(RpmOstreeImporter) unpacker =
+    rpmostree_importer_new_take_fd (&fd, ostreerepo, pkg, flags,
+                                    self->sepolicy, error);
+  if (!unpacker)
+    return FALSE;
+
+  if (rojig_xattrs)
+    {
+      g_assert (!self->sepolicy);
+      rpmostree_importer_set_rojig_mode (unpacker, self->rojig_xattr_table, rojig_xattrs);
+    }
+
+  rpmostree_importer_run_async (unpacker, cancellable, on_async_import_done, self);
+
+  return TRUE;
+}
+
+/* First function run on mainloop, and called after completion as well. Ensures
+ * that we have a bounded number of tasks concurrently executing until
+ * finishing.
+ */
+static gboolean
+async_imports_mainctx_iter (gpointer user_data)
+{
+  RpmOstreeContext *self = user_data;
+
+  while (self->async_index < self->pkgs_to_import->len &&
+         self->n_async_running < self->n_async_max)
+    {
+      DnfPackage *pkg = self->pkgs_to_import->pdata[self->async_index];
+      if (!import_one_package (self, pkg, self->async_cancellable, &self->async_error))
+        {
+          if (self->async_cancellable)
+            g_cancellable_cancel (self->async_cancellable);
```

Comment:
Minor: it looks like `g_cancellable_cancel` already supports `NULL` values, like the `GError` family.